test/test_rpc.py

```diff
@@ -281,7 +304,146 @@ def test_cause_service_error(self):
+        # Hacky strategy to achieve this to, in a loop:
+        # - call async connect()
+        # - after some delay, call async close()
+        # - with each loop, the delay gets slightly longer
+        # - break out of loop loop once the delay is long enough that the
+        #   connect() is completing before we ever call close()
+        delay_increment_sec = 0.005
+        stop_after_n_successful_connections = 2
+
+        delay_sec = 0.0
+        successful_connections = 0
+        while successful_connections < stop_after_n_successful_connections:
+            # not using helper _connect() call because it blocks until async connect() completes
+            self.lifecycle_handler = ConnectionLifecycleHandler(self._on_handler_freakout)
+            connect_future = self.connection.connect(self.lifecycle_handler)
+
+            if delay_sec > 0.0:
+                sleep(delay_sec)
+            close_future = self.connection.close()
+
+            # wait for connect and close to complete
+            connect_exception = connect_future.exception(TIMEOUT)
+            close_exception = close_future.exception(TIMEOUT)
+
+            # close should have been clean
+            self.assertIsNone(close_exception)
+
+            # connect might have succeeded, or might have failed,
+            # depending on the timing of this thread's close() call
+            if connect_exception:
+                self.assertIsInstance(connect_exception, ConnectionClosedError)
+                # lifecycle handlers should NOT fire if connect setup failed
+                # wait a tiny bit to be 100% sure these never fire
+                self.assertFalse(self.lifecycle_handler.connect_event.wait(0.1))
+                self.assertFalse(self.lifecycle_handler.disconnect_event.wait(0.1))
+            else:
+                self.assertTrue(self.lifecycle_handler.connect_event.wait(TIMEOUT))
+                self.assertTrue(self.lifecycle_handler.disconnect_event.wait(TIMEOUT))
+                successful_connections += 1
+
+            delay_sec += delay_increment_sec
+            self._assertNoHandlerFreakout()
+
+    def test_operation_response_completes_if_connection_closed(self):
+        # test that response future completes if connection is closed
+        # before actual response is received.
+
+        # this test is timing dependent, the response might actually
+        # come on another thread before this thread can close the connection,
+        # so run it in a loop till we get the timing we want
+        closed_before_response = False
+        while not closed_before_response:
+            self._connect()
+
+            stream_handler = StreamHandler(self._on_handler_freakout)
+            operation = self.echo_client.new_echo_stream_messages(stream_handler)
+            stream_handler.operation = operation
+
+            operation.activate(model.EchoStreamingRequest())
+            close_future = self.connection.close()
+
+            try:
```

Comment:
Do we want to put a cap on this? How often does it fail to be ordered right? How many times is it going to hang in CI?

Reply:
will do